http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/MultiplexedChannelRecord.java

```diff
@@ -31,165 +27,194 @@
+     * Delivers the exception to all registered child channels, and prohibits new streams being created on this connection.
      */
-    private void createChildChannel(Promise<Channel> channelPromise) {
-        doInEventLoop(connection.eventLoop(), () -> createChildChannel0(channelPromise), channelPromise);
-    }
+    void closeChildren(Throwable t) {
+        doInEventLoop(connection.eventLoop(), () -> {
+            if (state == RecordState.CLOSED) {
+                return;
+            }
+            state = RecordState.CLOSED;
 
-    private void createChildChannel0(Promise<Channel> channelPromise) {
-        if (goAway) {
-            channelPromise.tryFailure(new IOException("No streams are available on this connection."));
-        } else {
-            // Once protocol future is notified then parent pipeline is configured and ready to go
-            connection.attr(PROTOCOL_FUTURE).get()
-                      .whenComplete(asyncPromiseNotifyingBiConsumer(bootstrapChildChannel(), channelPromise));
-        }
+            // Create a copy of the children to close, because fireExceptionCaught may remove from the childChannels.
+            List<Http2StreamChannel> childrenToClose = new ArrayList<>(childChannels.values());
+            for (Channel childChannel : childrenToClose) {
+                childChannel.pipeline().fireExceptionCaught(t);
+            }
+        });
     }
 
-    /**
-     * Bootstraps the child stream channel and notifies the Promise on success or failure.
-     *
-     * @return BiConsumer that will bootstrap the child channel.
-     */
-    private BiConsumer<Protocol, Promise<Channel>> bootstrapChildChannel() {
-        return (s, p) -> new Http2StreamChannelBootstrap(connection)
-            .open()
-            .addListener((GenericFutureListener<Future<Http2StreamChannel>>) future -> {
-                if (future.isSuccess()) {
-                    Http2StreamChannel channel = future.getNow();
-                    childChannels.put(channel.id(), channel);
-                } else {
-                    if (!connection.isActive()) {
-                        channelReleaser.accept(connection, this);
-                    }
-                    availableStreams.incrementAndGet();
-                }
-            })
-            .addListener(promiseNotifyingListener(p));
+    public void closeAndReleaseChild(Channel childChannel) {
+        childChannel.close();
+        doInEventLoop(connection.eventLoop(), () -> {
+            childChannels.remove(childChannel.id());
+            releaseReservation();
+        });
     }
 
-    void release(Channel channel) {
-        availableStreams.incrementAndGet();
-        childChannels.remove(channel.id());
+    public Channel getConnection() {
+        return connection;
     }
 
-    boolean reusable() {
-        return !isPingInflight() && availableStreams.get() > 0 && !goAway;
-    }
+    public boolean reserveStream() {
```

Comment:
It's a public method. It's a way to atomically check stream availability AND reserve a stream, so that we don't have to synchronize access to the record.